File: GameServer/GameServerApp/GameServerApp/Proto/CreateProto/GameLevelSuccessResponseProto.cs
//===================================================
//作    者：xxx
//创建时间：2019-05-18 15:13:23
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary> 服务器返回关卡胜利协议 </summary>
public struct GameLevelSuccessResponseProto : IProto
{
    public ushort ProtoCode { get { return 50402; } }

    /// <summary> 是否胜利 </summary>
    public bool IsSuccess;
    /// <summary> 消息码 </summary>
    public int MsgCode;

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteBool(IsSuccess);
            if(!IsSuccess)
            {
                ms.WriteInt(MsgCode);
            }
            return ms.ToArray();
        }
    }

    public static GameLevelSuccessResponseProto GetProto(byte[] buffer)
    {
        GameLevelSuccessResponseProto proto = new GameLevelSuccessResponseProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.IsSuccess = ms.ReadBool();
            if(!proto.IsSuccess)
            {
                proto.MsgCode = ms.ReadInt();
            }
        }
        return proto;
    }
}

File: MMORPG_2/Assets/Script/Data/Proto/CreateLuaProto/GameLevelSuccessResponseProto.lua
--服务器返回关卡胜利协议
GameLevelSuccessResponseProto = { ProtoCode = 50402, IsSuccess = false, MsgCode = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
GameLevelSuccessResponseProto.__index = GameLevelSuccessResponseProto;

function GameLevelSuccessResponseProto.New()
    local self = { }; --初始化self
    setmetatable(self, GameLevelSuccessResponseProto); --将self的元表设定为Class
    return self;
end


--发送协议
function GameLevelSuccessResponseProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteBool(proto.IsSuccess);
    if(not proto.IsSuccess) then
        ms:WriteInt(MsgCode);
    end

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function GameLevelSuccessResponseProto.GetProto(buffer)

    local proto = GameLevelSuccessResponseProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.IsSuccess = ms:ReadBool();
    if(not proto.IsSuccess) then
        proto.MsgCode = ms:ReadInt();
    end

    ms:Dispose();
    return proto;
end